File: shared/src/main/java/com/runtracker/shared/data/model/CustomRunWorkout.kt
package com.runtracker.shared.data.model

import androidx.room.Entity
import androidx.room.PrimaryKey
import androidx.room.TypeConverters
import com.runtracker.shared.data.db.Converters

/**
 * A user-created custom running workout template
 */
@Entity(tableName = "custom_run_workouts")
@TypeConverters(Converters::class)
data class CustomRunWorkout(
    @PrimaryKey(autoGenerate = true)
    val id: Long = 0,
    val name: String,
    val description: String = "",
    val phases: List<WorkoutPhase> = emptyList(),
    val estimatedDurationMinutes: Int = 0,
    val estimatedDistanceMeters: Double = 0.0,
    val difficulty: RunDifficulty = RunDifficulty.MODERATE,
    val category: WorkoutCategory = WorkoutCategory.GENERAL,
    val isFavorite: Boolean = false,
    val timesUsed: Int = 0,
    val lastUsed: Long? = null,
    val createdAt: Long = System.currentTimeMillis()
) {
    val totalDurationSeconds: Int
        get() = phases.sumOf { phase -> 
            (phase.durationSeconds ?: 0) * phase.repetitions 
        }
    
    val formattedDuration: String
        get() {
            val totalSeconds = totalDurationSeconds
            val hours = totalSeconds / 3600
            val minutes = (totalSeconds % 3600) / 60
            return if (hours > 0) "${hours}h ${minutes}m" else "${minutes}m"
        }
}

/**
 * A single phase within a custom workout (warmup, interval, recovery, cooldown)
 */
data class WorkoutPhase(
    val id: String = java.util.UUID.randomUUID().toString(),
    val type: PhaseType,
    val name: String = "",
    val durationSeconds: Int? = null,
    val distanceMeters: Double? = null,
    val targetPaceMin: Double? = null,      // seconds per km
    val targetPaceMax: Double? = null,      // seconds per km
    val targetHeartRateMin: Int? = null,
    val targetHeartRateMax: Int? = null,
    val targetHeartRateZone: HeartRateZone? = null,
    val effort: EffortLevel = EffortLevel.MODERATE,
    val repetitions: Int = 1,
    val orderIndex: Int = 0,
    val notes: String? = null
) {
    val durationFormatted: String
        get() {
            val seconds = durationSeconds ?: return "--"
            val mins = seconds / 60
            val secs = seconds % 60
            return if (mins > 0) {
                if (secs > 0) "${mins}m ${secs}s" else "${mins}m"
            } else {
                "${secs}s"
            }
        }
    
    val distanceFormatted: String
        get() {
            val meters = distanceMeters ?: return "--"
            return if (meters >= 1000) {
                "%.1f km".format(meters / 1000)
            } else {
                "${meters.toInt()}m"
            }
        }
    
    val paceRangeFormatted: String
        get() {
            val min = targetPaceMin ?: return "--"
            val max = targetPaceMax
            val minFormatted = formatPace(min)
            return if (max != null && max != min) {
                "$minFormatted - ${formatPace(max)}"
            } else {
                minFormatted
            }
        }
    
    private fun formatPace(secondsPerKm: Double): String {
        val mins = (secondsPerKm / 60).toInt()
        val secs = (secondsPerKm % 60).toInt()
        return "$mins:${secs.toString().padStart(2, '0')}"
    }
}

enum class PhaseType {
    WARMUP,
    WORK,
    RECOVERY,
    REST,
    COOLDOWN,
    STRIDE,         // Short acceleration
    FLOAT,          // Easy running between hard efforts
    SURGE,          // Brief speed increase
    TEMPO,          // Sustained threshold effort
    SPRINT          // All-out effort
}

enum class EffortLevel {
    VERY_EASY,      // Can hold a conversation easily
    EASY,           // Conversational pace
    MODERATE,       // Can speak in sentences
    HARD,           // Can only speak a few words
    VERY_HARD,      // Cannot speak
    ALL_OUT         // Maximum effort
}

enum class RunDifficulty {
    EASY,
    MODERATE,
    HARD,
    VERY_HARD
}

enum class WorkoutCategory {
    GENERAL,
    SPEED,
    ENDURANCE,
    STRENGTH,
    RECOVERY,
    RACE_PREP,
    CUSTOM
}

/**
 * A user-created custom training plan
 */
@Entity(tableName = "custom_training_plans")
@TypeConverters(Converters::class)
data class CustomTrainingPlan(
    @PrimaryKey(autoGenerate = true)
    val id: Long = 0,
    val name: String,
    val description: String = "",
    val goalType: GoalType = GoalType.CUSTOM,
    val targetRaceDistance: Double? = null,     // meters
    val targetRaceTime: Long? = null,           // milliseconds
    val targetRaceDate: Long? = null,
    val durationWeeks: Int = 12,
    val startDate: Long? = null,
    val weeks: List<PlanWeek> = emptyList(),
    val currentWeek: Int = 1,
    val isActive: Boolean = false,
    val isFavorite: Boolean = false,
    val createdAt: Long = System.currentTimeMillis(),
    val lastModified: Long = System.currentTimeMillis()
) {
    val totalWorkouts: Int
        get() = weeks.sumOf { it.workouts.size }
    
    val completedWorkouts: Int
        get() = weeks.sumOf { week -> week.workouts.count { it.isCompleted } }
    
    val progressPercent: Float
        get() = if (totalWorkouts > 0) completedWorkouts.toFloat() / totalWorkouts else 0f
}

/**
 * A week within a custom training plan
 */
data class PlanWeek(
    val weekNumber: Int,
    val name: String = "Week $weekNumber",
    val description: String = "",
    val weekType: WeekType = WeekType.BUILD,
    val workouts: List<PlanWorkout> = emptyList(),
    val totalDistanceMeters: Double = 0.0,
    val totalDurationMinutes: Int = 0
)

/**
 * A scheduled workout within a plan week
 */
data class PlanWorkout(
    val id: String = java.util.UUID.randomUUID().toString(),
    val dayOfWeek: Int,                         // 1 = Monday, 7 = Sunday
    val workoutType: WorkoutType,
    val customWorkoutId: Long? = null,          // Reference to CustomRunWorkout if custom
    val name: String = "",
    val description: String = "",
    val targetDistanceMeters: Double? = null,
    val targetDurationMinutes: Int? = null,
    val targetPaceMin: Double? = null,
    val targetPaceMax: Double? = null,
    val intervals: List<Interval>? = null,
    val isCompleted: Boolean = false,
    val completedRunId: Long? = null,
    val completedDate: Long? = null,
    val notes: String? = null
)

enum class WeekType {
    BASE,           // Foundation building
    BUILD,          // Increasing intensity/volume
    PEAK,           // Highest training load
    TAPER,          // Reducing load before race
    RECOVERY,       // Easy week for adaptation
    RACE            // Race week
}

/**
 * Pre-built workout templates for common workouts
 */
object RunWorkoutTemplates {
    
    fun getSpeedWorkouts(): List<CustomRunWorkout> = listOf(
        CustomRunWorkout(
            name = "Classic 400m Repeats",
            description = "8x400m with 200m recovery jog",
            category = WorkoutCategory.SPEED,
            difficulty = RunDifficulty.HARD,
            phases = listOf(
                WorkoutPhase(type = PhaseType.WARMUP, durationSeconds = 600, effort = EffortLevel.EASY, name = "Easy jog"),
                WorkoutPhase(type = PhaseType.STRIDE, durationSeconds = 20, effort = EffortLevel.HARD, repetitions = 4, name = "Strides"),
                WorkoutPhase(type = PhaseType.WORK, distanceMeters = 400.0, effort = EffortLevel.VERY_HARD, repetitions = 8, name = "400m repeat"),
                WorkoutPhase(type = PhaseType.RECOVERY, distanceMeters = 200.0, effort = EffortLevel.EASY, repetitions = 8, name = "Recovery jog"),
                WorkoutPhase(type = PhaseType.COOLDOWN, durationSeconds = 600, effort = EffortLevel.VERY_EASY, name = "Cool down")
            )
        ),
        CustomRunWorkout(
            name = "800m Repeats",
            description = "6x800m at 5K pace with 400m recovery",
            category = WorkoutCategory.SPEED,
            difficulty = RunDifficulty.HARD,
            phases = listOf(
                WorkoutPhase(type = PhaseType.WARMUP, durationSeconds = 600, effort = EffortLevel.EASY),
                WorkoutPhase(type = PhaseType.WORK, distanceMeters = 800.0, effort = EffortLevel.VERY_HARD, repetitions = 6),
                WorkoutPhase(type = PhaseType.RECOVERY, distanceMeters = 400.0, effort = EffortLevel.EASY, repetitions = 6),
                WorkoutPhase(type = PhaseType.COOLDOWN, durationSeconds = 600, effort = EffortLevel.VERY_EASY)
            )
        ),
        CustomRunWorkout(
            name = "Mile Repeats",
            description = "4x1 mile at threshold pace with 3 min rest",
            category = WorkoutCategory.SPEED,
            difficulty = RunDifficulty.VERY_HARD,
            phases = listOf(
                WorkoutPhase(type = PhaseType.WARMUP, durationSeconds = 600, effort = EffortLevel.EASY),
                WorkoutPhase(type = PhaseType.WORK, distanceMeters = 1609.0, effort = EffortLevel.HARD, repetitions = 4),
                WorkoutPhase(type = PhaseType.REST, durationSeconds = 180, effort = EffortLevel.VERY_EASY, repetitions = 4),
                WorkoutPhase(type = PhaseType.COOLDOWN, durationSeconds = 600, effort = EffortLevel.VERY_EASY)
            )
        ),
        CustomRunWorkout(
            name = "Yasso 800s",
            description = "Marathon predictor: 10x800m at goal marathon time",
            category = WorkoutCategory.SPEED,
            difficulty = RunDifficulty.VERY_HARD,
            phases = listOf(
                WorkoutPhase(type = PhaseType.WARMUP, durationSeconds = 600, effort = EffortLevel.EASY),
                WorkoutPhase(type = PhaseType.WORK, distanceMeters = 800.0, effort = EffortLevel.HARD, repetitions = 10),
                WorkoutPhase(type = PhaseType.RECOVERY, distanceMeters = 400.0, effort = EffortLevel.EASY, repetitions = 10),
                WorkoutPhase(type = PhaseType.COOLDOWN, durationSeconds = 600, effort = EffortLevel.VERY_EASY)
            )
        )
    )
    
    fun getLadderWorkouts(): List<CustomRunWorkout> = listOf(
        CustomRunWorkout(
            name = "Classic Ladder",
            description = "200-400-600-800-600-400-200m pyramid",
            category = WorkoutCategory.SPEED,
            difficulty = RunDifficulty.HARD,
            phases = listOf(
                WorkoutPhase(type = PhaseType.WARMUP, durationSeconds = 600, effort = EffortLevel.EASY),
                WorkoutPhase(type = PhaseType.WORK, distanceMeters = 200.0, effort = EffortLevel.VERY_HARD),
                WorkoutPhase(type = PhaseType.RECOVERY, durationSeconds = 60, effort = EffortLevel.EASY),
                WorkoutPhase(type = PhaseType.WORK, distanceMeters = 400.0, effort = EffortLevel.VERY_HARD),
                WorkoutPhase(type = PhaseType.RECOVERY, durationSeconds = 90, effort = EffortLevel.EASY),
                WorkoutPhase(type = PhaseType.WORK, distanceMeters = 600.0, effort = EffortLevel.VERY_HARD),
                WorkoutPhase(type = PhaseType.RECOVERY, durationSeconds = 120, effort = EffortLevel.EASY),
                WorkoutPhase(type = PhaseType.WORK, distanceMeters = 800.0, effort = EffortLevel.VERY_HARD),
                WorkoutPhase(type = PhaseType.RECOVERY, durationSeconds = 120, effort = EffortLevel.EASY),
                WorkoutPhase(type = PhaseType.WORK, distanceMeters = 600.0, effort = EffortLevel.VERY_HARD),
                WorkoutPhase(type = PhaseType.RECOVERY, durationSeconds = 90, effort = EffortLevel.EASY),
                WorkoutPhase(type = PhaseType.WORK, distanceMeters = 400.0, effort = EffortLevel.VERY_HARD),
                WorkoutPhase(type = PhaseType.RECOVERY, durationSeconds = 60, effort = EffortLevel.EASY),
                WorkoutPhase(type = PhaseType.WORK, distanceMeters = 200.0, effort = EffortLevel.VERY_HARD),
                WorkoutPhase(type = PhaseType.COOLDOWN, durationSeconds = 600, effort = EffortLevel.VERY_EASY)
            )
        )
    )
    
    fun getTempoWorkouts(): List<CustomRunWorkout> = listOf(
        CustomRunWorkout(
            name = "Classic Tempo",
            description = "20 min at threshold pace",
            category = WorkoutCategory.ENDURANCE,
            difficulty = RunDifficulty.MODERATE,
            phases = listOf(
                WorkoutPhase(type = PhaseType.WARMUP, durationSeconds = 600, effort = EffortLevel.EASY),
                WorkoutPhase(type = PhaseType.TEMPO, durationSeconds = 1200, effort = EffortLevel.HARD),
                WorkoutPhase(type = PhaseType.COOLDOWN, durationSeconds = 600, effort = EffortLevel.VERY_EASY)
            )
        ),
        CustomRunWorkout(
            name = "Cruise Intervals",
            description = "4x5 min at threshold with 1 min recovery",
            category = WorkoutCategory.ENDURANCE,
            difficulty = RunDifficulty.HARD,
            phases = listOf(
                WorkoutPhase(type = PhaseType.WARMUP, durationSeconds = 600, effort = EffortLevel.EASY),
                WorkoutPhase(type = PhaseType.TEMPO, durationSeconds = 300, effort = EffortLevel.HARD, repetitions = 4),
                WorkoutPhase(type = PhaseType.RECOVERY, durationSeconds = 60, effort = EffortLevel.EASY, repetitions = 4),
                WorkoutPhase(type = PhaseType.COOLDOWN, durationSeconds = 600, effort = EffortLevel.VERY_EASY)
            )
        ),
        CustomRunWorkout(
            name = "Progression Run",
            description = "Start easy, finish at tempo pace",
            category = WorkoutCategory.ENDURANCE,
            difficulty = RunDifficulty.MODERATE,
            phases = listOf(
                WorkoutPhase(type = PhaseType.WARMUP, durationSeconds = 300, effort = EffortLevel.VERY_EASY, name = "Very easy start"),
                WorkoutPhase(type = PhaseType.WORK, durationSeconds = 600, effort = EffortLevel.EASY, name = "Easy pace"),
                WorkoutPhase(type = PhaseType.WORK, durationSeconds = 600, effort = EffortLevel.MODERATE, name = "Moderate pace"),
                WorkoutPhase(type = PhaseType.TEMPO, durationSeconds = 600, effort = EffortLevel.HARD, name = "Tempo finish"),
                WorkoutPhase(type = PhaseType.COOLDOWN, durationSeconds = 300, effort = EffortLevel.VERY_EASY)
            )
        )
    )
    
    fun getHillWorkouts(): List<CustomRunWorkout> = listOf(
        CustomRunWorkout(
            name = "Hill Repeats",
            description = "8x60s uphill with jog down recovery",
            category = WorkoutCategory.STRENGTH,
            difficulty = RunDifficulty.HARD,
            phases = listOf(
                WorkoutPhase(type = PhaseType.WARMUP, durationSeconds = 600, effort = EffortLevel.EASY),
                WorkoutPhase(type = PhaseType.WORK, durationSeconds = 60, effort = EffortLevel.VERY_HARD, repetitions = 8, name = "Hill sprint"),
                WorkoutPhase(type = PhaseType.RECOVERY, durationSeconds = 90, effort = EffortLevel.VERY_EASY, repetitions = 8, name = "Jog down"),
                WorkoutPhase(type = PhaseType.COOLDOWN, durationSeconds = 600, effort = EffortLevel.VERY_EASY)
            )
        ),
        CustomRunWorkout(
            name = "Hill Sprints",
            description = "10x20s all-out hill sprints",
            category = WorkoutCategory.STRENGTH,
            difficulty = RunDifficulty.VERY_HARD,
            phases = listOf(
                WorkoutPhase(type = PhaseType.WARMUP, durationSeconds = 600, effort = EffortLevel.EASY),
                WorkoutPhase(type = PhaseType.SPRINT, durationSeconds = 20, effort = EffortLevel.ALL_OUT, repetitions = 10, name = "Sprint"),
                WorkoutPhase(type = PhaseType.RECOVERY, durationSeconds = 120, effort = EffortLevel.VERY_EASY, repetitions = 10, name = "Full recovery"),
                WorkoutPhase(type = PhaseType.COOLDOWN, durationSeconds = 600, effort = EffortLevel.VERY_EASY)
            )
        )
    )
    
    fun getFartlekWorkouts(): List<CustomRunWorkout> = listOf(
        CustomRunWorkout(
            name = "Classic Fartlek",
            description = "Unstructured speed play with varied efforts",
            category = WorkoutCategory.SPEED,
            difficulty = RunDifficulty.MODERATE,
            phases = listOf(
                WorkoutPhase(type = PhaseType.WARMUP, durationSeconds = 600, effort = EffortLevel.EASY),
                WorkoutPhase(type = PhaseType.SURGE, durationSeconds = 60, effort = EffortLevel.HARD, repetitions = 6),
                WorkoutPhase(type = PhaseType.FLOAT, durationSeconds = 120, effort = EffortLevel.EASY, repetitions = 6),
                WorkoutPhase(type = PhaseType.COOLDOWN, durationSeconds = 600, effort = EffortLevel.VERY_EASY)
            )
        ),
        CustomRunWorkout(
            name = "Mona Fartlek",
            description = "2x(90s hard, 90s easy, 60s hard, 60s easy, 30s hard, 30s easy)",
            category = WorkoutCategory.SPEED,
            difficulty = RunDifficulty.HARD,
            phases = listOf(
                WorkoutPhase(type = PhaseType.WARMUP, durationSeconds = 600, effort = EffortLevel.EASY),
                WorkoutPhase(type = PhaseType.WORK, durationSeconds = 90, effort = EffortLevel.HARD, repetitions = 2),
                WorkoutPhase(type = PhaseType.FLOAT, durationSeconds = 90, effort = EffortLevel.EASY, repetitions = 2),
                WorkoutPhase(type = PhaseType.WORK, durationSeconds = 60, effort = EffortLevel.HARD, repetitions = 2),
                WorkoutPhase(type = PhaseType.FLOAT, durationSeconds = 60, effort = EffortLevel.EASY, repetitions = 2),
                WorkoutPhase(type = PhaseType.WORK, durationSeconds = 30, effort = EffortLevel.VERY_HARD, repetitions = 2),
                WorkoutPhase(type = PhaseType.FLOAT, durationSeconds = 30, effort = EffortLevel.EASY, repetitions = 2),
                WorkoutPhase(type = PhaseType.COOLDOWN, durationSeconds = 600, effort = EffortLevel.VERY_EASY)
            )
        )
    )
    
    fun getRecoveryWorkouts(): List<CustomRunWorkout> = listOf(
        CustomRunWorkout(
            name = "Recovery Run",
            description = "Easy 30 min recovery jog",
            category = WorkoutCategory.RECOVERY,
            difficulty = RunDifficulty.EASY,
            phases = listOf(
                WorkoutPhase(type = PhaseType.WARMUP, durationSeconds = 300, effort = EffortLevel.VERY_EASY),
                WorkoutPhase(type = PhaseType.WORK, durationSeconds = 1500, effort = EffortLevel.EASY),
                WorkoutPhase(type = PhaseType.COOLDOWN, durationSeconds = 300, effort = EffortLevel.VERY_EASY)
            )
        ),
        CustomRunWorkout(
            name = "Shake Out Run",
            description = "Short easy run before or after race",
            category = WorkoutCategory.RECOVERY,
            difficulty = RunDifficulty.EASY,
            phases = listOf(
                WorkoutPhase(type = PhaseType.WORK, durationSeconds = 900, effort = EffortLevel.VERY_EASY),
                WorkoutPhase(type = PhaseType.STRIDE, durationSeconds = 15, effort = EffortLevel.MODERATE, repetitions = 4),
                WorkoutPhase(type = PhaseType.WORK, durationSeconds = 300, effort = EffortLevel.VERY_EASY)
            )
        )
    )
    
    fun getAllTemplates(): List<CustomRunWorkout> = 
        getSpeedWorkouts() + getLadderWorkouts() + getTempoWorkouts() + 
        getHillWorkouts() + getFartlekWorkouts() + getRecoveryWorkouts()
}

File: iosApp/GoSteady/Sources/Models/CustomRunWorkout.swift
import Foundation
import SwiftData

@Model
final class CustomRunWorkout {
    @Attribute(.unique) var id: UUID
    var name: String
    var workoutDescription: String
    var phases: [WorkoutPhase]
    var estimatedDurationMinutes: Int
    var estimatedDistanceMeters: Double
    var difficulty: RunDifficulty
    var category: WorkoutCategory
    var isFavorite: Bool
    var timesUsed: Int
    var lastUsed: Date?
    var createdAt: Date

    var totalDurationSeconds: Int {
        phases.reduce(0) { $0 + ($1.durationSeconds ?? 0) * $1.repetitions }
    }

    var formattedDuration: String {
        let totalSeconds = totalDurationSeconds
        let hours = totalSeconds / 3600
        let minutes = (totalSeconds % 3600) / 60
        return hours > 0 ? "\(hours)h \(minutes)m" : "\(minutes)m"
    }

    init(
        id: UUID = UUID(),
        name: String,
        workoutDescription: String = "",
        phases: [WorkoutPhase] = [],
        estimatedDurationMinutes: Int = 0,
        estimatedDistanceMeters: Double = 0.0,
        difficulty: RunDifficulty = .moderate,
        category: WorkoutCategory = .general,
        isFavorite: Bool = false,
        timesUsed: Int = 0,
        lastUsed: Date? = nil,
        createdAt: Date = Date()
    ) {
        self.id = id
        self.name = name
        self.workoutDescription = workoutDescription
        self.phases = phases
        self.estimatedDurationMinutes = estimatedDurationMinutes
        self.estimatedDistanceMeters = estimatedDistanceMeters
        self.difficulty = difficulty
        self.category = category
        self.isFavorite = isFavorite
        self.timesUsed = timesUsed
        self.lastUsed = lastUsed
        self.createdAt = createdAt
    }
}

struct WorkoutPhase: Codable, Hashable, Identifiable {
    var id: String
    var type: PhaseType
    var name: String
    var durationSeconds: Int?
    var distanceMeters: Double?
    var targetPaceMin: Double?
    var targetPaceMax: Double?
    var targetHeartRateMin: Int?
    var targetHeartRateMax: Int?
    var targetHeartRateZone: HeartRateZone?
    var effort: EffortLevel
    var repetitions: Int
    var orderIndex: Int
    var notes: String?

    var durationFormatted: String {
        guard let seconds = durationSeconds else { return "--" }
        let mins = seconds / 60
        let secs = seconds % 60
        if mins > 0 {
            return secs > 0 ? "\(mins)m \(secs)s" : "\(mins)m"
        } else {
            return "\(secs)s"
        }
    }

    var distanceFormatted: String {
        guard let meters = distanceMeters else { return "--" }
        if meters >= 1000 {
            return String(format: "%.1f km", meters / 1000)
        } else {
            return "\(Int(meters))m"
        }
    }

    var paceRangeFormatted: String {
        guard let minPace = targetPaceMin else { return "--" }
        let minFormatted = formatPace(minPace)
        if let maxPace = targetPaceMax, maxPace != minPace {
            return "\(minFormatted) - \(formatPace(maxPace))"
        }
        return minFormatted
    }

    private func formatPace(_ secondsPerKm: Double) -> String {
        let mins = Int(secondsPerKm / 60)
        let secs = Int(secondsPerKm.truncatingRemainder(dividingBy: 60))
        return "\(mins):\(String(format: "%02d", secs))"
    }

    init(
        id: String = UUID().uuidString,
        type: PhaseType,
        name: String = "",
        durationSeconds: Int? = nil,
        distanceMeters: Double? = nil,
        targetPaceMin: Double? = nil,
        targetPaceMax: Double? = nil,
        targetHeartRateMin: Int? = nil,
        targetHeartRateMax: Int? = nil,
        targetHeartRateZone: HeartRateZone? = nil,
        effort: EffortLevel = .moderate,
        repetitions: Int = 1,
        orderIndex: Int = 0,
        notes: String? = nil
    ) {
        self.id = id
        self.type = type
        self.name = name
        self.durationSeconds = durationSeconds
        self.distanceMeters = distanceMeters
        self.targetPaceMin = targetPaceMin
        self.targetPaceMax = targetPaceMax
        self.targetHeartRateMin = targetHeartRateMin
        self.targetHeartRateMax = targetHeartRateMax
        self.targetHeartRateZone = targetHeartRateZone
        self.effort = effort
        self.repetitions = repetitions
        self.orderIndex = orderIndex
        self.notes = notes
    }
}

enum PhaseType: String, Codable, CaseIterable {
    case warmup = "WARMUP"
    case work = "WORK"
    case recovery = "RECOVERY"
    case rest = "REST"
    case cooldown = "COOLDOWN"
    case stride = "STRIDE"
    case float = "FLOAT"
    case surge = "SURGE"
    case tempo = "TEMPO"
    case sprint = "SPRINT"
}

enum EffortLevel: String, Codable, CaseIterable {
    case veryEasy = "VERY_EASY"
    case easy = "EASY"
    case moderate = "MODERATE"
    case hard = "HARD"
    case veryHard = "VERY_HARD"
    case allOut = "ALL_OUT"
}

enum RunDifficulty: String, Codable, CaseIterable {
    case easy = "EASY"
    case moderate = "MODERATE"
    case hard = "HARD"
    case veryHard = "VERY_HARD"
}

enum WorkoutCategory: String, Codable, CaseIterable {
    case general = "GENERAL"
    case speed = "SPEED"
    case endurance = "ENDURANCE"
    case strength = "STRENGTH"
    case recovery = "RECOVERY"
    case racePrep = "RACE_PREP"
    case custom = "CUSTOM"
}

@Model
final class CustomTrainingPlan {
    @Attribute(.unique) var id: UUID
    var name: String
    var planDescription: String
    var goalType: GoalType
    var targetRaceDistance: Double?
    var targetRaceTime: Int64?
    var targetRaceDate: Date?
    var durationWeeks: Int
    var startDate: Date?
    var weeks: [PlanWeek]
    var currentWeek: Int
    var isActive: Bool
    var isFavorite: Bool
    var createdAt: Date
    var lastModified: Date

    var totalWorkouts: Int {
        weeks.reduce(0) { $0 + $1.workouts.count }
    }

    var completedWorkouts: Int {
        weeks.reduce(0) { sum, week in sum + week.workouts.filter(\.isCompleted).count }
    }

    var progressPercent: Float {
        totalWorkouts > 0 ? Float(completedWorkouts) / Float(totalWorkouts) : 0
    }

    init(
        id: UUID = UUID(),
        name: String,
        planDescription: String = "",
        goalType: GoalType = .custom,
        targetRaceDistance: Double? = nil,
        targetRaceTime: Int64? = nil,
        targetRaceDate: Date? = nil,
        durationWeeks: Int = 12,
        startDate: Date? = nil,
        weeks: [PlanWeek] = [],
        currentWeek: Int = 1,
        isActive: Bool = false,
        isFavorite: Bool = false,
        createdAt: Date = Date(),
        lastModified: Date = Date()
    ) {
        self.id = id
        self.name = name
        self.planDescription = planDescription
        self.goalType = goalType
        self.targetRaceDistance = targetRaceDistance
        self.targetRaceTime = targetRaceTime
        self.targetRaceDate = targetRaceDate
        self.durationWeeks = durationWeeks
        self.startDate = startDate
        self.weeks = weeks
        self.currentWeek = currentWeek
        self.isActive = isActive
        self.isFavorite = isFavorite
        self.createdAt = createdAt
        self.lastModified = lastModified
    }
}

struct PlanWeek: Codable, Hashable {
    var weekNumber: Int
    var name: String
    var weekDescription: String
    var weekType: WeekType
    var workouts: [PlanWorkout]
    var totalDistanceMeters: Double
    var totalDurationMinutes: Int

    init(
        weekNumber: Int,
        name: String? = nil,
        weekDescription: String = "",
        weekType: WeekType = .build,
        workouts: [PlanWorkout] = [],
        totalDistanceMeters: Double = 0.0,
        totalDurationMinutes: Int = 0
    ) {
        self.weekNumber = weekNumber
        self.name = name ?? "Week \(weekNumber)"
        self.weekDescription = weekDescription
        self.weekType = weekType
        self.workouts = workouts
        self.totalDistanceMeters = totalDistanceMeters
        self.totalDurationMinutes = totalDurationMinutes
    }
}

struct PlanWorkout: Codable, Hashable, Identifiable {
    var id: String
    var dayOfWeek: Int
    var workoutType: WorkoutType
    var customWorkoutId: UUID?
    var name: String
    var workoutDescription: String
    var targetDistanceMeters: Double?
    var targetDurationMinutes: Int?
    var targetPaceMin: Double?
    var targetPaceMax: Double?
    var intervals: [Interval]?
    var isCompleted: Bool
    var completedRunId: UUID?
    var completedDate: Date?
    var notes: String?

    init(
        id: String = UUID().uuidString,
        dayOfWeek: Int,
        workoutType: WorkoutType,
        customWorkoutId: UUID? = nil,
        name: String = "",
        workoutDescription: String = "",
        targetDistanceMeters: Double? = nil,
        targetDurationMinutes: Int? = nil,
        targetPaceMin: Double? = nil,
        targetPaceMax: Double? = nil,
        intervals: [Interval]? = nil,
        isCompleted: Bool = false,
        completedRunId: UUID? = nil,
        completedDate: Date? = nil,
        notes: String? = nil
    ) {
        self.id = id
        self.dayOfWeek = dayOfWeek
        self.workoutType = workoutType
        self.customWorkoutId = customWorkoutId
        self.name = name
        self.workoutDescription = workoutDescription
        self.targetDistanceMeters = targetDistanceMeters
        self.targetDurationMinutes = targetDurationMinutes
        self.targetPaceMin = targetPaceMin
        self.targetPaceMax = targetPaceMax
        self.intervals = intervals
        self.isCompleted = isCompleted
        self.completedRunId = completedRunId
        self.completedDate = completedDate
        self.notes = notes
    }
}

enum WeekType: String, Codable, CaseIterable {
    case base = "BASE"
    case build = "BUILD"
    case peak = "PEAK"
    case taper = "TAPER"
    case recovery = "RECOVERY"
    case race = "RACE"
}
